File: ios/TKMFontLoader.m
// Copyright 2018 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "TKMFontLoader.h"

#import <Foundation/Foundation.h>
#import <CoreText/CoreText.h>

#import "UserDefaults.h"

static dispatch_once_t sOnceToken;
static NSArray<TKMFont*> *sLoadedFonts;
static NSPredicate * sEnabledPredicate;
static NSString *const kFontDirectory = @"fonts";

NSString *LoadFontFromUrl(NSURL *url) {
  NSFileManager *fileManager = [NSFileManager defaultManager];
  NSData *data = [fileManager contentsAtPath: [url path]];
  
  CFErrorRef error;
  CGDataProviderRef provider = CGDataProviderCreateWithCFData((CFDataRef)data);
  CGFontRef font = CGFontCreateWithDataProvider(provider);
  if (! CTFontManagerRegisterGraphicsFont(font, &error)) {
    CFStringRef errorDescription = CFErrorCopyDescription(error);
    NSLog(@"Failed to load font: %@", errorDescription);
    CFRelease(errorDescription);
    CFRelease(font);
    CFRelease(provider);
    return nil;
  }
  CFStringRef fontName = CGFontCopyFullName(font);
  
  CFRelease(font);
  CFRelease(provider);
  return (__bridge NSString *) fontName;
}

void EnsureInitialized() {
  dispatch_once(&sOnceToken, ^{
    sEnabledPredicate = [NSPredicate predicateWithFormat:@"enabled == true"];
    
    NSBundle *mainBundle = [NSBundle mainBundle];
    NSArray<NSURL*> *urls = [mainBundle URLsForResourcesWithExtension:nil subdirectory:kFontDirectory];
    
    NSMutableArray<TKMFont*> *fonts = [NSMutableArray array];
    
    NSArray<TKMFont*> *fontsFromDefaults = UserDefaults.usedFonts;
    
    for (NSURL *url in urls) {
      TKMFont *newFont = [[TKMFont alloc] init];
      newFont.fontName = LoadFontFromUrl(url);
      if (newFont.fontName == nil) {
        continue;
      }
      
      NSPredicate *fontNamePredicate = [NSPredicate predicateWithFormat: @"fontName == %@", newFont.fontName];
      NSArray<TKMFont*> *filteredArray = [fontsFromDefaults filteredArrayUsingPredicate:fontNamePredicate];
      newFont.enabled = filteredArray == nil || filteredArray.firstObject.enabled;
      
      [fonts addObject:newFont];
    }
    
    sLoadedFonts = fonts;
  });
}


@implementation TKMFontLoader

+ (NSArray<TKMFont*> *)getLoadedFonts {
  EnsureInitialized();
  return sLoadedFonts;
}

+ (NSArray<TKMFont*> *)getEnabledFonts {
  return [[TKMFontLoader getLoadedFonts] filteredArrayUsingPredicate:sEnabledPredicate];
}

+ (BOOL)font:(TKMFont*)font canRender:(NSString*)text {
  CTFontRef fontRef = CTFontCreateWithName((CFStringRef)font.fontName, 0.0, NULL);
  
  NSUInteger count = text.length;
  unichar characters[count];
  CGGlyph glyphs[count];
  [text getCharacters:characters range:NSMakeRange(0, count)];
  
  BOOL canRender = CTFontGetGlyphsForCharacters(fontRef, characters, glyphs, count);
  CFRelease(fontRef);
  
  return canRender;
}

+ (TKMFont*)getRandomFontToRender:(NSString*)text {
  NSPredicate *fontPredicate = [NSPredicate predicateWithBlock:^BOOL(TKMFont *font, NSDictionary *_) {
    return font.enabled && [TKMFontLoader font:font canRender:text];
  }];
  
  NSArray<TKMFont*> *availableFonts = [[TKMFontLoader getLoadedFonts] filteredArrayUsingPredicate:fontPredicate];
  NSUInteger random = arc4random_uniform((uint32_t)[availableFonts count]);
  return availableFonts[random];
}

+ (void) saveToUserDefaults {
  UserDefaults.usedFonts = sLoadedFonts;
}

@end

@implementation TKMFont

- (instancetype)initWithFontName:(NSString*)fontName enabled:(BOOL)enabled{
  self = [super init];
  if (self) {
    _fontName = fontName;
    _enabled = enabled;
  }
  return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
  NSString *fontName = (NSString*)[aDecoder decodeObjectForKey:@"fontName"];
  BOOL enabled = [aDecoder decodeBoolForKey:@"enabled"];
  
  return [self initWithFontName:fontName enabled:enabled];
}

- (void)encodeWithCoder:(NSCoder *)coder {
  [coder encodeObject:_fontName forKey:@"fontName"];
  [coder encodeBool:_enabled forKey:@"enabled"];
}

@end

File: ios/TKMFontLoader.swift
// Copyright 2020 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import CoreText
import Foundation

struct FontDefinition {
  var fontName: String
  var fileName: String
  var displayName: String
  var sizeBytes: Int64
  init(_ fontName: String, _ fileName: String, _ displayName: String, _ sizeBytes: Int64) {
    self.fontName = fontName
    self.fileName = fileName
    self.displayName = displayName
    self.sizeBytes = sizeBytes
  }
}

@objcMembers class TKMFont: NSObject {
  static let fontPreviewText = "色は匂へど散りぬるを我が世誰ぞ常ならん有為の奥山今日越えて浅き夢見じ酔ひもせず"
  static let fontDefinitions: [FontDefinition] = [
    FontDefinition("ArmedBanana", "armed-banana.ttf", "Armed Banana", 3_298_116),
    FontDefinition("darts font", "darts-font.woff", "Darts", 1_349_440),
    FontDefinition("Hosofuwafont", "hoso-fuwa.ttf", "Hoso Fuwa", 5_910_760),
    FontDefinition("nagayama_kai", "nagayama-kai.otf", "Nagayama Kai calligraphy", 15_576_732),
    FontDefinition("santyoume-font", "san-chou-me.ttf", "San Chou Me", 4_428_896),
  ]
  static func loadFont(path: String) -> Bool {
    guard let data = FileManager.default.contents(atPath: path) else { return false }
    let provider = CGDataProvider(data: NSData(data: data))!
    let font = CGFont(provider)!
    return CTFontManagerRegisterGraphicsFont(font, nil)
  }

  func canRenderText(fontName: String, text: String) -> Bool {
    let fontRef = CTFontCreateWithName(NSString(string: fontName), 0.0, nil)
    let count = text.count, characters = text.utf16.map { $0 }
    let glyphs = UnsafeMutablePointer<CGGlyph>.allocate(capacity: count)
    let canRender = CTFontGetGlyphsForCharacters(fontRef, characters, glyphs, count)
    if canRender {
      // Check every glyph has a path.
      for i in 0 ..< count {
        guard CTFontCreatePathForGlyph(fontRef, glyphs[i], nil) != nil else { return false }
      }
    }
    return canRender
  }

  var definition: FontDefinition
  var fontName: String { definition.fontName }
  var fileName: String { definition.fileName }
  var displayName: String { definition.displayName }
  var sizeBytes: Int64 { definition.sizeBytes }
  var available: Bool = false

  init(definition: FontDefinition) {
    self.definition = definition
    super.init()
    reload()
  }

  func reload() {
    if available { return }
    if UIFont.fontNames(forFamilyName: fontName).count > 0 {
      available = true
      return
    }

    // Try to load a built-in font first.
    if let path = Bundle.main.path(forResource: "fonts/\(fileName)", ofType: nil) {
      if TKMFont.loadFont(path: path) {
        available = true
        return
      }
    }

    // Try to load the downloaded font.
    let path = "\(TKMFontLoader.cacheDirectoryPath)/\(fileName)"
    NSLog("Loading font \(path)")
    if TKMFont.loadFont(path: path) {
      available = true
      return
    }
  }

  func didDelete() { available = false }
  func loadScreenshot() -> UIImage! { UIImage(named: fontName) }
}

@objcMembers class TKMFontLoader: NSObject {
  static var cacheDirectoryPath: String {
    "\(NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0])/fonts"
  }

  let allFonts: [TKMFont]
  override init() {
    var allFonts: [TKMFont] = []
    for definition in TKMFont.fontDefinitions {
      allFonts.append(TKMFont(definition: definition))
    }
    self.allFonts = allFonts
    super.init()
  }

  func font(name fileName: String) -> TKMFont? {
    for font in allFonts {
      if font.fileName == fileName { return font }
    }
    return nil
  }
}
